File: 2024/03/main.ts
import { sum, product } from '../shared/math.ts';
import { run } from '../shared/runtime.ts';

export type ParseOutput = ReturnType<typeof parse>;

export function parse(input: string) {
  const lines = input
    .matchAll(/(mul|do|don't)\((?:(\d{1,3}),(\d{1,3}))?\)/g)
    .toArray()
    .map(([, instruction, ...args]) => ({ instruction, args: args.map(Number) as [number, number] }));
  return lines;
}

export function solvePart1(input: ParseOutput) {
  return sumFilteredProducts(input, ({ instruction }) => instruction === 'mul');
}

export function solvePart2(input: ParseOutput) {
  let enabled = true;
  return sumFilteredProducts(
    input,
    ({ instruction }) => (instruction === 'mul' && enabled) || ((enabled = instruction === 'do') && false)
  );
}

function sumFilteredProducts(input: ParseOutput, filter: (instruction: ParseOutput[number]) => boolean) {
  return input
    .filter(filter)
    .map(({ args }) => product(...args))
    .reduce(sum);
}

if (import.meta.main) {
  run(parse, solvePart1, solvePart2);
}

File: 2024/03/main.go
package main

import (
	"regexp"
	"strconv"

	"github.com/janvoracek/advent-of-code/aoc/2024/shared"
)

type ParseOutput = string

func parse(input string) ParseOutput {
	return input
}

func solvePart1(input ParseOutput) any {
	re := regexp.MustCompile(`mul\((\d{1,3}),(\d{1,3})\)`)
	matches := re.FindAllStringSubmatch(input, -1)
	sum := 0
	for _, match := range matches {
		sum += parseInt(match[1]) * parseInt(match[2])
	}
	return sum
}

func solvePart2(input ParseOutput) any {
	re := regexp.MustCompile(`(?s)don't\(\).+?(do\(\)|$)`)
	return solvePart1(re.ReplaceAllString(input, ""))
}

func parseInt(s string) int {
	return shared.HandleError(strconv.Atoi(s))
}

func main() {
	shared.Run(parse, solvePart1, solvePart2)
}
